### ISSUE TITLE
Windows compatibility and conceptual considerations

### ISSUE THREAD

**Issue by mbuceta** [davidsantiago/pathetic#2]
IMHO there is a conceptual problem with the library. But first, the concrete issues:

The library won't even load on Windows because

``` clojure
(re-pattern File/separator)
```

is illegal in Windows where path separator is "\".  "\" must be escaped inside a regex. But, that could be easily solved doing

``` clojure
(re-pattern (Pattern/quote File/separator))
```

Once that is done, you'd need to correct a couple of places where separator and separator-pattern are confused in the code, since you can't treat them as equal once the pattern is escaped.

Now to the (IMHO) conceptual problem: The library intends to provide Unix like path operations, `File/separator` shouldn't appear anywhere in the code. Users from any OS should **always** use `"/"` with this library, the fact that there is a different separator in their systems has  nothing to do with it, the library only takes care of unix-like paths only.

For the same reason, arguments can't never be instances of `java.io.File`, since those are system dependent (in windows `(File. "/")` gets converted to `(File. "\\")`). Those tests using `File` as argument are condemned to fail in Windows.

In conclusion, why so many `(File.)` and `File/separator` everywhere? According to what I understand there shouldn't be any, and that way the library would work as intended in any platform.

Go with

``` clojure
(def ^{:private true} default-separator "/")
```

and remove any tests and documentation mentioning `File` instances as arguments.

Sorry about the length and abstract nature, hope it helps

**Comment by davidsantiago** [comment]
You know what's funny, is the reason I added the separator args and what not was so that, even if it's not officially supported by the library, Windows users who have limited needs, say only relative paths, might be able to slide something by and get some use out of it anyways. Of course, I don't know any Windows users, and certainly no Windows users who work with JVM languages, so that was a shot in the dark. I certainly don't know the Windows programming environment, so if you think that is not useful for Windows users, I'm thrilled to not try to support that use case.

Support for File objects is useful, though. I wasn't aware of that behavior from File objects, so I may create some special handling for File objects, to ensure that they are parsed with the path separator used in that object. Would you be available to test these changes?

**Comment by paraseba** [comment]
Hi, I'm the author of the ticket (I was in someone else's machine). I'm not a Windows user either, I just built an app using https://github.com/cemerick/url (which uses pathetic) and sent it to a friend who uses windows. It didn't load. That's why I came back hear to try to understand what was going on.

I think that for windows users it is better to have a library that is not very useful but that at least doesn't break any other lib or app that uses it. And windows users have a perfectly good reason to want this library: handling URL paths.

What I'd do? Declare the library as "/" only, drop every File use. File is an abstraction over path strings representations in different OS. Your library doesn't need that, your library handles one and only one representation of paths, using "/". If your application is using File, is because it wants to abstract filesystem paths, independently of your OS. We can't help you, we just handle a particular representation that unix and URLs use.

Other option would be to say  "this library helps you handle filesystem paths in a OS independent way". That would require some work, it would be useful, but it won't help with URLs. I'm pretty sure this is not what you wanted when you created the library.

Is this making any sense?

**Comment by davidsantiago** [comment]
You are making sense, and I'm with you. This library is and always has been about unix-style path handling, and I definitely want it to work in Windows for that purpose, even if it doesn't serve Windows paths. 

Where we are having a disagreement is about how much more use you could get out of it. The library parses paths into an internal representation, and if you have simple needs, say a tool that does things with files in a local and relative subdirectory structure, it could be useful for that in a cross-platform way. My mistake was not understanding the intricacies of File objects when used on Windows (as well as my inability to test such code). The difficulty with this use case appears to be entirely in the movement to/from the OS-specific world and the abstract path structure. Once you are working with an abstract list of path segments, that can still be useful to have. One could think of various ways to make parse-path and render-path into protocols, allowing a third party to write implementations that could work on Windows, or something like that. I am not interested in doing so, and thought that File provided more or less that functionality. 

HOWEVER, you have convinced me to do more or less what you say simply because I cannot test the code on Windows. If I can't test it, it's not worth writing. So I'm going to remove the use of File objects from the code. They will still be somewhat usable in the sense that one could call .toString on them to get their path, if the coder knows that their code is not meant to run on Windows. 

  David

**Comment by davidsantiago** [comment]
I just released version 0.4.0, which removes configurable path separators and any use of java.io.File. Hopefully this will work for you. I will ping cemerick about updating the url lib next time he comes on IRC.

**Comment by paraseba** [comment]
Awesome! I agree with your previous comment, it is in fact something useful to have implemented in a OS independent way, and it shouldn't be hard to do it. But you should provide a mechanism to say "do the following manipulation using this separator". Otherwise, windows users wouldn't be able to use the library to handle URLs, it couldn't be used for cemerick's url for instance. Maybe something like:

``` clojure
(with-path-separator "/"
  ...)
```

I'll try the new version as soon as I can put my hand on a Windows machine, and will report back.

Thanks for taking the time to do this!
